config-controller/pkg/client/client.go

```diff
@@ -246,36 +247,56 @@ func (c *client) CreatePolicy(ctx context.Context, policy *storage.Policy) (*sto
 		return nil, errors.Wrap(err, fmt.Sprintf("Failed to POST policy '%s'", policy.Name))
 	}
 
-	c.cache[createdPolicy.Name] = createdPolicy
+	c.policyObjectCache[createdPolicy.GetId()] = createdPolicy
+	c.policyNameToIDCache[createdPolicy.GetName()] = createdPolicy.GetId()
 
 	return createdPolicy, nil
 }
 
 func (c *client) UpdatePolicy(ctx context.Context, policy *storage.Policy) error {
 	getLogger(ctx).Info("PUT", "policyName", policy.Name)
-	err := c.svc.PutPolicy(ctx, policy)
 
+	existingPolicy := c.policyObjectCache[policy.GetId()]
+	//Do not update/overwrite if name collision with imperative policies
+	if id, ok := c.policyNameToIDCache[policy.GetName()]; ok {
+		if existingPolicy.GetSource() != storage.PolicySource_DECLARATIVE {
+			return errors.New(fmt.Sprintf("Failed to PUT polict '%s' due to name collision with existing policy '%s'. You may either rename this policy in your source control,"+
+				"or edit the name of the policy managed by Central, and retry the operation.",
+				policy.GetName(), id))
+		}
+	}
+
+	// update policy on central
+	err := c.svc.PutPolicy(ctx, policy)
 	if err != nil {
 		return errors.Wrap(err, "Failed to PUT policy")
 	}
-
-	c.cache[policy.Name] = policy
+	// update caches
+	c.policyObjectCache[policy.GetId()] = policy
+	if existingPolicy.GetName() != policy.GetName() {
+		delete(c.policyNameToIDCache, existingPolicy.GetName())
+		c.policyNameToIDCache[policy.GetName()] = policy.GetId()
+	}
 
 	return nil
 }
 
 func (c *client) DeletePolicy(ctx context.Context, name string) error {
 	getLogger(ctx).Info("DELETE", "policyName", name)
-	policy := c.cache[name]
-
+	policyID, ok := c.policyNameToIDCache[name]
```

Comment:
why not just pass in ID instead of name to avoid this lookup?

Reply:
Can do, although I'd be relying on a status field which may not exist for any arcane/error path situations so preferred the name, at the cost of an extra lookup. Thoughts?

Reply:
I don't feel too strongly. You can leave it as is